demos/multi_material/benchmarks/crameri_2012.py

```diff
@@ -0,0 +1,173 @@
+from functools import partial
+
+import initial_signed_distance as isd
+
+import gadopt as ga
+
+
+class Mantle(ga.AbstractMaterial):
```

Comment:
I'm wondering what the best structure here is. So it seems you are sub-classing specific classes for each specific Material from AbstractMaterial, and then you're only using these as a class, they're never substantiated, but the way you use them as a class, they really just act like an object of which you call some methods. So why not just:
```
Mantle = Material(B=None, RaB=None, density=3300, viscosity=1e21, thermal_expansion=1, ...)
```
where Material is a concrete (data) class that just sets a bunch of attributes - in fact you could use an actual [datclass](https://docs.python.org/3/library/dataclasses.html) for that.
Of course that would make them attributes rather than methods, but I wasn't immediately clear what the use case for them being methods was. In the AbstractMaterial docstring you mention "to allow for temporal and spatial dependency" but I don't really follow tbh. The only example where it's actually used seems to be the viscosity of the lithosphere in the Schmallholz case, but there it's just referring to some extra attributes that have just been added on the fly, and really don't need to be a part of the class/object. In the end of the day you still end up with an expression that is always the same, so this could just be set as an attribute. If in some specific case you really want some code to "calculate" what the expression should be (and this calculation is not just a one off that you could have just done beforehand), you could always subclass Material, and change sub-class the attribute to become a property.

The other slight smell about AbstractMaterial is that it's defined in gadopt, but there's no associated code in gadopt. The only place we actually _use_ the materials is in the run_benchmark demo code. An abstract class is like a contract that defines what data and methods will be available but the meaning of that data is always intimately tied with the code that is written to handle it. So for instance you seem to assign specific meaning to B or RaB having the value None, but that logic lives in that bit of demo code.
Note I'm not so much concerned about your Simulation class and its class methods, that seems a kind of nice and reasonable way to avoid code duplication between the different cases, where you just stuff the specific bits in a class, and that's just the way you have chosen to set up the demos. My concern is more with the structure of AbstractMaterial which would be part of gadopt/ and also as a by-effect forces uses to actually know what a classmethod is (we have very intelligent users but not all of them are CS wizards).

Reply:
I agree with the current structure not being ideal, but I fail to see how a `dataclass` could improve things. I can see two issues here. First, it would be nice to be able to provide an integer, float, or tuple if the value of a material property is constant. However, we also have to account for the fact that the material property could be `Callable`. For example, viscosity may depend on velocity and temperature, which evolve from one timestep to another, and thus it needs to be evaluated at each timestep. Indeed, this involves a function call, and objects of `float` type, for example, do not define `__call__`. Say we enforce material properties as callables. Then how different is it to subclass an `ABC` from writing functions and providing them to a `dataclass`? In the current format, all material properties are methods of a class, and each class corresponds to a material, so there is a sense of structure. Happy to have more thoughts.

Reply:
How I imagine it could look like as a `dataclass`.
```python
@dataclass(kw_only=True)
class Material:
    """Dataclass to specify relevant material properties that affect simulation's
    evolution. Each material property is provided as a class attribute. All physical
    properties should be dimensional."""

    B: int | float | Callable | None = None  # Buoyancy number
    RaB: int | float | Callable | None = None  # Compositional Rayleigh number
    density: int | float | Callable | None = None
    viscosity: int | float | Callable | None = 1  # Dynamic viscosity
    thermal_expansion: int | float | Callable | None = 1  # Volumetric expansion
    thermal_conductivity: int | float | Callable | None = 1
    specific_heat_capacity: int | float | Callable | None = 1  # At constant pressure
    internal_heating_rate: int | float | Callable | None = 0  # Per unit mass

    def __post_init__(self):
        self.thermal_diffusivity = (
            self.thermal_conductivity / self.density / self.specific_heat_capacity
        )
```

Reply:
How would you implement the lithospheric viscosity from the Schmalholz benchmark here?

Reply:
Since you've effectively defined an implicit contract where dynamic viscosity is _always_ provided a velocity argument, you could provide that as an argument.

I've provided a demo of this: https://github.com/g-adopt/g-adopt/commit/b9127724bd979ef67a4acee2c05e2e8fab28cb36. Because none of the other properties take an argument, they can be turned into a function as a `@property`, in an overriding class.

Reply:
Thanks for the suggestion! I have tried to combine elements from Steph's suggestion, my example above, and your example. Let me know what you think of what I just pushed.